src/memory/memory.cpp

```diff
@@ -0,0 +1,256 @@
+};
+
+#ifndef GETNAME
+#define GETNAME(x) case x: return #x;
+#endif
+
+// Stringify RMM error code.
+const char * rmmGetErrorString(rmmError_t errcode) {
+  switch (errcode) {
+    // There must be one entry per enum values in gdf_error.
+    GETNAME(RMM_SUCCESS)
+    GETNAME(RMM_ERROR_CUDA_ERROR)
+    GETNAME(RMM_ERROR_INVALID_ARGUMENT)
+    GETNAME(RMM_ERROR_NOT_INITIALIZED)
+    GETNAME(RMM_ERROR_OUT_OF_MEMORY)
+    GETNAME(RMM_ERROR_UNKNOWN)
+    GETNAME(RMM_ERROR_IO)
+    default:
+        // This means we are missing an entry above for a rmmError_t value.
+        return "Internal error. Unknown error code.";
+  }
+}
+
+// Initialize memory manager state and storage.
+rmmError_t rmmInitialize()
+{
+    theManager = new rmm::Manager;
+#ifndef RMM_USE_CUDAMALLOC
+    cnmemDevice_t dev;
+    RMM_CHECK_CUDA( cudaGetDevice(&(dev.device)) );
+    dev.size = 16800000000;
+    dev.numStreams = 1;
+    cudaStream_t streams[1]; streams[0] = 0;
+    dev.streams = streams;
+    dev.streamSizes = 0;
+    RMM_CHECK_CNMEM( cnmemInit(1, &dev, 0) );    
+#endif
+    return RMM_SUCCESS;
+}
+
+// Shutdown memory manager.
+rmmError_t rmmFinalize()
+{
+#ifndef RMM_USE_CUDAMALLOC
+    RMM_CHECK_CNMEM( cnmemFinalize() );
+#endif
+    delete theManager;
+    return RMM_SUCCESS;
+}
+ 
+// Allocate memory and return a pointer to device memory. 
+rmmError_t rmmAlloc(void **ptr, size_t size, cudaStream_t stream)
+{
+    LogIt log(rmm::Logger::Alloc, size, stream);
+
+    if (!ptr && !size) {
+        return RMM_SUCCESS;
+    }
+    
+    if (!ptr) 
+        return RMM_ERROR_INVALID_ARGUMENT;
+
+#ifdef RMM_USE_CUDAMALLOC
+    RMM_CHECK_CUDA(cudaMalloc(ptr, size));
+#else
+    RMM_CHECK( theManager->registerStream(stream) );
+    RMM_CHECK_CNMEM( cnmemMalloc(ptr, size, stream) );
+#endif
+    log.setPointer(*ptr);
+    return RMM_SUCCESS;
+}
+
+// Reallocate device memory block to new size and recycle any remaining memory.
+rmmError_t rmmRealloc(void **ptr, size_t new_size, cudaStream_t stream)
+{
+    LogIt log(rmm::Logger::Realloc, new_size, stream);
+
+	if (!ptr && !new_size) {
+        return RMM_SUCCESS;
+    }
+
+    if (!ptr) 
+    	return RMM_ERROR_INVALID_ARGUMENT;
+
+#ifdef RMM_USE_CUDAMALLOC
+    RMM_CHECK_CUDA(cudaFree(*ptr));
+	RMM_CHECK_CUDA(cudaMalloc(ptr, new_size));
+#else
+    RMM_CHECK( theManager->registerStream(stream) );
+    RMM_CHECK_CNMEM( cnmemFree(*ptr, stream) );
+    RMM_CHECK_CNMEM( cnmemMalloc(ptr, new_size, stream) );
+#endif
+    return RMM_SUCCESS;
+}
+
+// Release device memory and recycle the associated memory.
+rmmError_t rmmFree(void *ptr, cudaStream_t stream)
+{
+    LogIt log(rmm::Logger::Free, ptr, 0, stream);
+#ifdef RMM_USE_CUDAMALLOC
+	RMM_CHECK_CUDA(cudaFree(ptr));
+#else
+    RMM_CHECK_CNMEM( cnmemFree(ptr, stream) );
+#endif
+	return RMM_SUCCESS;
+}
+
+// Get the offset of ptr from its base allocation
+rmmError_t rmmGetAllocationOffset(ptrdiff_t *offset, void *ptr, cudaStream_t stream)
+{
+#ifdef RMM_USE_CUDAMALLOC
```

Comment:
This doesn't seem correct, should it be `#ifndef`?

Reply:
Yeah, I think @nsakharnykh is right.

Reply:
This actually popped into my head right after I pushed but I was boarding a flight. That's what I get for programming before boarding.